ksp_plugin/pile_up.cpp

```diff
@@ -419,44 +426,153 @@ void PileUp::DeformPileUpIfNeeded(Instant const& t) {
+
+  Rotation<ApparentPileUp, EquivalentRigidPileUp>
+      apparent_pile_up_equivalent_rotation =
+          trivial_rotations
+              ? Rotation<ApparentPileUp, EquivalentRigidPileUp>::Identity()
+              : Rotation<ApparentPileUp, EquivalentRigidPileUp>(
+                    apparent_correction_axis,
+                    L̂_apparent,
+                    Commutator(apparent_correction_axis, L̂_apparent));
+  Rotation<NonRotatingPileUp, EquivalentRigidPileUp>
+      actual_pile_up_equivalent_rotation =
+          trivial_rotations
+              ? Rotation<NonRotatingPileUp, EquivalentRigidPileUp>::Identity()
+              : Rotation<NonRotatingPileUp, EquivalentRigidPileUp>(
+                    actual_correction_axis,
+                    L̂_actual,
+                    Commutator(actual_correction_axis, L̂_actual));
+  Rotation<ApparentPileUp, NonRotatingPileUp> const
+      tentative_attitude_correction =
+          actual_pile_up_equivalent_rotation.Inverse() *
+          apparent_pile_up_equivalent_rotation;
+
   // The angular velocity of a rigid body with the inertia and angular momentum
   // of the apparent parts.
   auto const apparent_equivalent_angular_velocity =
       apparent_angular_momentum / inertia_tensor;
   // The angular velocity of a rigid body with the inertia of the apparent
   // parts, and the angular momentum of the pile up.
-  auto const actual_equivalent_angular_velocity =
-      angular_momentum_ /
-      Identity<ApparentPileUp, NonRotatingPileUp>()(inertia_tensor);
+  auto actual_equivalent_angular_velocity =
+      angular_momentum_ / tentative_attitude_correction(inertia_tensor);
+
+  // So far we have only dealt with the essential singularity by removing it.
+  // We now need to deal with its neighbourhood, by ensuring that the tentative
+  // correction is not too big compared to the angular velocities involved.
+  Quaternion const q = tentative_attitude_correction.quaternion();
+  // α is the angle of the tentative attitude correction.
+  Angle const α =
+      2 * quantities::ArcTan(q.imaginary_part().Norm(), q.real_part());
+  AngularFrequency const ω = std::min(apparent_equivalent_angular_velocity.Norm(),
+                          actual_equivalent_angular_velocity.Norm());
+  Time const Δt = t - psychohistory_->back().time;
+  if (thresholding && α > ω * Δt) {
+    // The attitude correction is too large.  Preserve attitude.
+    apparent_pile_up_equivalent_rotation =
+        Rotation<ApparentPileUp, EquivalentRigidPileUp>::Identity();
+    actual_pile_up_equivalent_rotation =
+        Rotation<NonRotatingPileUp, EquivalentRigidPileUp>::Identity();
+    actual_equivalent_angular_velocity =
+        angular_momentum_ /
+        Identity<ApparentPileUp, NonRotatingPileUp>()(inertia_tensor);
+  }
+
+  bool const correcting_orientation =
+      apparent_pile_up_equivalent_rotation.quaternion() != Quaternion(1) ||
+      actual_pile_up_equivalent_rotation.quaternion() != Quaternion(1);
```

Comment:
We should also be able to write `rotation != Identity()`.  More stuff for the above-mentioned issue.

Reply:
Added to #2534.